From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
Date: Sat, 10 Oct 2026 12:35:54 +0200
Subject: [PATCH] Build argv and call main without newlib

build_argv came from devkitPPC's newlib and __crtmain from its crtmain.o,
both now live here.
---
 libogc/argv.c | 30 +++++++++++++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/libogc/argv.c b/libogc/argv.c
index 6dea388..c63cbb2 100644
--- a/libogc/argv.c
+++ b/libogc/argv.c
@@ -25,11 +25,29 @@ distribution.
 
 -------------------------------------------------------------*/
 #include "gctypes.h"
+#include <stdlib.h>
 #include <string.h>
 
 extern u8 __Arena1Lo[];
 extern char *__argvArena1Lo;
-void build_argv (struct __argv* argstruct );
+
+// Splits the null separated command line into an argv array, this used to live in newlib
+static void build_argv(struct __argv* argstruct)
+{
+	char** argv = (char**)(((int)argstruct->commandLine + argstruct->length + 3) & ~3);
+	char* data = argstruct->commandLine;
+	int len = argstruct->length;
+	int argc = 0;
+	while (len > 0) {
+		argv[argc++] = data;
+		int size = strlen(data) + 1;
+		data += size;
+		len -= size;
+	}
+	argstruct->argc = argc;
+	argstruct->argv = argv;
+	argstruct->endARGV = &argv[argc];
+}
 
 void __CheckARGV() {
 
@@ -48,3 +66,13 @@ void __CheckARGV() {
 	__argvArena1Lo = (char *)__system_argv->endARGV;	
 
 }
+
+// Entry point of the main thread, this used to live in devkitPPC's crtmain.o
+extern int main(int argc, char **argv);
+extern void SYS_PreMain(void);
+
+void __crtmain(void)
+{
+	SYS_PreMain();
+	exit(main(__system_argv->argc, __system_argv->argv));
+}
